Content.Server/Chat/V2/Systems/ChatAttemptHandlerSystem.cs

```diff
@@ -0,0 +1,235 @@
+        {
+            RaiseNetworkEvent(new RadioFailedEvent(ev.Sender, reason), args.SenderSession);
+
+            return;
+        }
+
+        _repo.Add(new RadioCreatedEvent(GetEntity(ev.Sender), SanitizeMessage(ev), ev.Channel));
+    }
+
+    private void HandleAttemptWhisper(AttemptWhisperEvent ev, EntitySessionEventArgs args)
+    {
+        if (!ValidateMessage(ev, args.SenderSession, out var reason, _whisperFailed, out CanWhisperComponent? comp))
+        {
+            RaiseNetworkEvent(new WhisperFailedEvent(ev.Sender, reason), args.SenderSession);
+
+            return;
+        }
+
+        _repo.Add(new WhisperCreatedEvent(GetEntity(ev.Sender), SanitizeMessage(ev), comp.MinRange, comp.MaxRange));
+    }
+
+    /// <summary>
+    /// Validates messages. Return true means the message is valid.
+    /// </summary>
+    private bool ValidateMessage<T>(T ev, ICommonSession player, out string reason) where T : ChatAttemptEvent
+    {
+        var entityUid = GetEntity(ev.Sender);
+
+        // This check is simple, so we don't need to raise an event for it.
+        if (player.AttachedEntity != null || player.AttachedEntity != entityUid)
+        {
+            reason = _entityNotOwnedBySender;
+
+            return false;
+        }
+
+        // Raise both the general and specific ChatValidationEvents. This allows for general
+        var validate = new ChatValidationEvent<ChatAttemptEvent>(ev);
+        RaiseLocalEvent(entityUid, validate);
+
+        if (validate.IsCancelled)
+        {
+            reason = validate.Reason;
+
+            return false;
+        }
+
+        var validateT = new ChatValidationEvent<T>(ev);
+        RaiseLocalEvent(entityUid, validateT);
+
+        if (validate.IsCancelled)
+        {
+            reason = validate.Reason;
+
+            return false;
+        }
+
+        reason = "";
+
+        return true;
+    }
+
+    /// <summary>
+    /// Validates messages which depend on a sentinel component to be legal.
+    /// </summary>
+    private bool ValidateMessage<T1, T2>(
+        T1 ev,
+        ICommonSession player,
+        out string reason,
+        string invalidCompReason,
+        [NotNullWhen(true)] out T2? comp
+    ) where T1 : ChatAttemptEvent where T2 : IComponent
+    {
+        comp = default;
+
+        var entityUid = GetEntity(ev.Sender);
+
+        if (!ValidateMessage(ev, player, out reason))
+        {
+            return false;
+        }
+
+        if (TryComp(entityUid, out comp))
+            return true;
+
+        reason = Loc.GetString(invalidCompReason);
+
+        return false;
+    }
+
+    /// <summary>
+    /// Sanitizes messages. The return string is sanitized.
+    /// </summary>
+    private string SanitizeMessage<T>(T evt) where T : ChatAttemptEvent
+    {
+        var sanitize = new ChatSanitizationEvent<T>(evt);
+        RaiseLocalEvent(ref sanitize);
```

Comment:
This is not struct, is there any need for ref here? It is misleading as it looks like struct ;( or am i mistaken?